asserts/batch.go

```diff
@@ -0,0 +1,212 @@
+			return err
+		}
+	}
+
+	return &Batch{
+		bs:            NewMemoryBackstore(),
+		inPrereqOrder: true, // empty list is trivially so
+		unsupported:   unsupported,
+	}
+}
+
+// Add one assertion to the batch.
+func (b *Batch) Add(a Assertion) error {
+	b.inPrereqOrder = false
+
+	if !a.SupportedFormat() {
+		err := &UnsupportedFormatError{Ref: a.Ref(), Format: a.Format()}
+		return b.unsupported(a.Ref(), err)
+	}
+	if err := b.bs.Put(a.Type(), a); err != nil {
+		if revErr, ok := err.(*RevisionError); ok {
+			if revErr.Current >= a.Revision() {
+				// we already got something more recent
+				return nil
+			}
+		}
+		return err
+	}
+	b.added = append(b.added, a)
+	return nil
+}
+
+// AddStream adds a stream of assertions to the batch.
+// Returns references to the assertions effectively added.
+func (b *Batch) AddStream(r io.Reader) ([]*Ref, error) {
+	b.inPrereqOrder = false
+
+	start := len(b.added)
+	dec := NewDecoder(r)
+	for {
+		a, err := dec.Decode()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if err := b.Add(a); err != nil {
+			return nil, err
+		}
+	}
+	added := b.added[start:]
+	if len(added) == 0 {
+		return nil, nil
+	}
+	refs := make([]*Ref, len(added))
+	for i, a := range added {
+		refs[i] = a.Ref()
+	}
+	return refs, nil
+}
+
+// Fetch adds to the batch by invoking fetching to drive an internal
+// Fetcher that was built with trustedDB and retrieve.
+func (b *Batch) Fetch(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), fetching func(Fetcher) error) error {
+	f := NewFetcher(trustedDB, retrieve, b.Add)
+	return fetching(f)
+}
+
+// Precheck pre-checks whether adding the batch of assertions to the
+// given assertion database should fully succeed.
+func (b *Batch) Precheck(db *Database) error {
+	db = db.WithStackedBackstore(NewMemoryBackstore())
+	return b.commitTo(db)
+}
+
+// CommitTo adds the batch of assertions to the given assertion database.
+func (b *Batch) CommitTo(db *Database) error {
+	return b.commitTo(db)
+}
+
+// commitTo does a best effort of adding all the batch assertions to
+// the target database.
+func (b *Batch) commitTo(db *Database) error {
+	if err := b.prereqSort(db); err != nil {
+		return err
+	}
+
+	// TODO: trigger w. caller a global sanity check if something is revoked
+	// (but try to save as much possible still),
+	// or err is a check error
+
+	var errs []error
+	for _, a := range b.added {
+		err := db.Add(a)
+		if IsUnaccceptedUpdate(err) {
+			// unsupported format case is handled before
+			// be idempotent
+			// system db has already the same or newer
+			continue
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) != 0 {
+		return &commitError{errs: errs}
+	}
+	return nil
+}
+
+func (b *Batch) prereqSort(db *Database) error {
+	if b.inPrereqOrder {
+		// nothing to do
+		return nil
+	}
+
+	// put in prereq order using a fetcher
+	inPrereqOrder := make([]Assertion, 0, len(b.added))
```

Comment:
This variable is named the same as the boolean flag which is mildly confusing. I would suggest giving it a new name, e.g. just `sorted`